actions/rules/action_state_to_metastate.go

```diff
@@ -0,0 +1,416 @@
+package rules
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+
+	uuid "github.com/satori/go.uuid"
+
+	"github.com/fabric8-services/fabric8-wit/application"
+	"github.com/fabric8-services/fabric8-wit/convert"
+	"github.com/fabric8-services/fabric8-wit/workitem"
+)
+
+type ActionStateToMetaState struct {
+	Db     application.DB
+	Ctx    context.Context
+	UserID *uuid.UUID
+}
+
+// make sure the rule is implementing the interface.
+var _ Action = ActionStateToMetaState{}
+
+func (act ActionStateToMetaState) contains(s []uuid.UUID, e uuid.UUID) bool {
+	for _, a := range s {
+		if a == e {
+			return true
+		}
+	}
+	return false
+}
+
+func (act ActionStateToMetaState) removeElement(s []uuid.UUID, e uuid.UUID) []uuid.UUID {
+	for idx, a := range s {
+		if a == e {
+			s = append(s[:idx], s[idx+1:]...)
+			// we don't return here as there may be multiple copies of e in s.
+		}
+	}
+	return s
+}
+
+func (act ActionStateToMetaState) difference(old []uuid.UUID, new []uuid.UUID) ([]uuid.UUID, []uuid.UUID) {
+	var added []uuid.UUID
+	var removed []uuid.UUID
+	// added in slice2
+	for _, elem := range new {
+		if !act.contains(old, elem) {
+			added = append(added, elem)
+		}
+	}
+	// removed in slice2
+	for _, elem := range old {
+		if !act.contains(new, elem) {
+			removed = append(removed, elem)
+		}
+	}
+	return added, removed
+}
+
+func (act ActionStateToMetaState) loadWorkItemBoardsBySpaceID(spaceID uuid.UUID) ([]*workitem.Board, error) {
+	if act.Ctx == nil {
+		return nil, errors.New("Context is nil")
+	}
+	if act.Db == nil {
+		return nil, errors.New("Database is nil")
+	}
+	space, err := act.Db.Spaces().Load(act.Ctx, spaceID)
+	if err != nil {
+		return nil, errors.New("Error loading space: " + err.Error())
+	}
+	boards, err := act.Db.Boards().List(act.Ctx, space.SpaceTemplateID)
+	if err != nil {
+		return nil, errors.New("Error loading work item type: " + err.Error())
+	}
+	return boards, nil
+}
+
+func (act ActionStateToMetaState) loadWorkItemTypeGroupsBySpaceID(spaceID uuid.UUID) ([]*workitem.WorkItemTypeGroup, error) {
+	if act.Ctx == nil {
+		return nil, errors.New("Context is nil")
+	}
+	if act.Db == nil {
```

Comment:
These checks should really be done at construction level. And maybe you want to hide the members behind a function `DB()` and `Ctx()` if you want to hide the implementation.

Reply:
How do I check this at creation and not use specific constructors? The creation of a new action instance should be as generic as possible to enable automatic creation based on a key and reflections (even if that is not done yet). I choose intentionally to create the rules using a literal as this gives this kind of flexibility we need. There are rules that will not need databases and/or contexts *or* need other resources (like a handle to a notification system).